libraries/state_history/include/eosio/state_history/log_catalog.hpp

```diff
@@ -0,0 +1,261 @@
+         retained_log_files.modify(log_it, [&](catalogued_log_file& clf) {
+            if(!clf.log)
+               clf.log.emplace(clf.path_and_basename, non_local_get_block_id);
+            clf.last_used_counter = ++global_used_counter;
+         });
+
+         const unsigned num_log_files_to_keep_open = 5;
+         if(retained_log_files.size() >= num_log_files_to_keep_open+1)
+            retained_log_files.get<by_mru>().modify(std::next(retained_log_files.get<by_mru>().begin(), num_log_files_to_keep_open), [](catalogued_log_file& clf) {
+               clf.log.reset();
+            });
+
+         return f(std::forward<state_history_log>(const_cast<state_history_log&>(*log_it->log)));
+      }
+      else
+         return f(std::forward<state_history_log>(*head_log));
+   }
+
+   void rotate_logs() {
+      const auto [begin, end] = head_log->block_range();
+      std::filesystem::path new_log_basenamepath = retained_dir / head_log_path_and_basename.stem();
+      new_log_basenamepath += "-" + std::to_string(begin) + "-" + std::to_string(end-1);
+      head_log.reset();
+      //try and make sure we don't leave head_log unset if something throws below. any throw below should cause the node to
+      // stop, but during teardown of everything something might access head_log which is assumed to always be set elsewhere
+      // TODO: only close the old log once the new log is opened to avoid problems when opening throws too
+      auto reopen_head_log = fc::make_scoped_exit([this]() {
+         open_head_log();
+      });
+
+      rename_bundle(head_log_path_and_basename, new_log_basenamepath);
+      retained_log_files.emplace(begin, end, new_log_basenamepath);
+
+      while(retained_log_files.size() > max_retained_files) {
+         const catalog_t::iterator it = retained_log_files.begin();
+         std::filesystem::path oldest_log_path_and_basename = it->path_and_basename;
+         if(archive_dir.empty()) {
+            std::filesystem::remove(oldest_log_path_and_basename.replace_extension("log"));
+            std::filesystem::remove(oldest_log_path_and_basename.replace_extension("index"));
+         } else {
+            rename_bundle(oldest_log_path_and_basename, archive_dir / oldest_log_path_and_basename.filename());
+         }
+         retained_log_files.erase(it);
+      }
+   }
+
+   void open_head_log(std::optional<state_history::prune_config> prune_config = std::nullopt) {
+      head_log.emplace(head_log_path_and_basename, non_local_get_block_id, prune_config);
+   }
+
+   static std::filesystem::path make_absolute_dir(const std::filesystem::path& base_dir, std::filesystem::path new_dir) {
+      if(new_dir.is_relative())
+         new_dir = base_dir / new_dir;
+
+      if(!std::filesystem::is_directory(new_dir))
+         std::filesystem::create_directories(new_dir);
+
+      return new_dir;
+   }
+
+   static void rename_if_not_exists(std::filesystem::path old_name, std::filesystem::path new_name) {
```

Comment:
The behavior seems a little odd to me too. But, this isn't new, it comes from the old implementation,
https://github.com/AntelopeIO/spring/blob/caa4780c9769f68cb63d04665f0927733ee848dd/libraries/chain/include/eosio/chain/log_catalog.hpp#L187-L195
I haven't really thought much of the behavior and what sort of problems it may cause.

Can you suggest a name you'd like?

Reply:
It is safe to keep it as is for now.